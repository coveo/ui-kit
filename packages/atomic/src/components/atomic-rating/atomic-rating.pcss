/**
 * @prop --atomic-rating-facet-icon-active-color: Color of the icon when active.
 */
.icon-active {
  color: var(--atomic-rating-facet-icon-active-color, #f4c315);
}

/**
 * @prop --atomic-rating-facet-icon-inactive-color: Color of the icon when inactive.
 */
.icon-inactive {
  color: var(--atomic-rating-facet-icon-inactive-color, var(--atomic-neutral));
}
