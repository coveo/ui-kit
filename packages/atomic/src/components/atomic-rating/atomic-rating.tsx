import {FunctionalComponent, h, VNode} from '@stencil/core';

interface RatingProps {
  numberOfTotalIcons: number;
  numberOfActiveIcons: number;
  icon: string;
  iconSize?: number;
}

export const Rating: FunctionalComponent<RatingProps> = (props) => {
  const width =
    ((props.numberOfActiveIcons / props.numberOfTotalIcons) * 100).toString() +
    '%';
  const iconSize = `${props.iconSize ?? 0.75}rem`;

  const renderIcon = (active: boolean) => {
    return (
      <atomic-icon
        icon={props.icon}
        class={`flex-shrink-0 ${active ? 'icon-active' : 'icon-inactive'}`}
        style={{width: iconSize, height: iconSize}}
      ></atomic-icon>
    );
  };

  const emptyIconDisplay = () => {
    const emptyIconDisplay: VNode[] = [];
    for (let i = 0; i < props.numberOfTotalIcons; i++) {
      emptyIconDisplay.push(renderIcon(false));
    }
    return emptyIconDisplay;
  };

  const filledIconDisplay = () => {
    const filledIconDisplay: VNode[] = [];
    for (let i = 0; i < props.numberOfTotalIcons; i++) {
      filledIconDisplay.push(renderIcon(true));
    }
    return filledIconDisplay;
  };

  return (
    <div class="inline-block relative left-0 top-0" part="value-rating">
      <div class="relative left-0 top-0 z-0 flex items-center gap-0.5 pt-0.5 pb-0.5">
        {emptyIconDisplay()}
      </div>
      <div
        class="absolute left-0 top-0 z-1 flex items-center gap-0.5 pt-0.5 pb-0.5 overflow-hidden"
        style={{width}}
      >
        {filledIconDisplay()}
      </div>
    </div>
  );
};
