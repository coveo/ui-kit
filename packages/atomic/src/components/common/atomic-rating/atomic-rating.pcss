/**
 * @prop --atomic-rating-icon-active-color: Color of the icon when active.
 */
.icon-active {
  color: var(--atomic-rating-icon-active-color, #f6ce3c);
}

/**
 * @prop --atomic-rating-icon-inactive-color: Color of the icon when inactive.
 */
.icon-inactive {
  color: var(--atomic-rating-icon-inactive-color, var(--atomic-neutral));
}

/**
 * @prop --atomic-rating-icon-outline: Color of the icon's outline.
 */
.icon-active,
.icon-inactive {
  & svg path {
    stroke: var(--atomic-rating-icon-outline, none);
  }
}
