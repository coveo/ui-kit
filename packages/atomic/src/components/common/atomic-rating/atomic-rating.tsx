import {FunctionalComponent, h, VNode} from '@stencil/core';
import {i18n} from 'i18next';

interface RatingProps {
  i18n: i18n;
  numberOfTotalIcons: number;
  numberOfActiveIcons: number;
  icon: string;
  iconSize?: number;
}

export const Rating: FunctionalComponent<RatingProps> = (props) => {
  const width =
    ((props.numberOfActiveIcons / props.numberOfTotalIcons) * 100).toString() +
    '%';
  const iconSize = `${props.iconSize ?? 0.75}rem`;

  const renderIcon = (active: boolean) => {
    return (
      <atomic-icon
        icon={props.icon}
        class={`shrink-0 ${active ? 'icon-active' : 'icon-inactive'}`}
        style={{width: iconSize, height: iconSize}}
      ></atomic-icon>
    );
  };

  const emptyIconDisplay = () => {
    const emptyIconDisplay: VNode[] = [];
    for (let i = 0; i < props.numberOfTotalIcons; i++) {
      emptyIconDisplay.push(renderIcon(false));
    }
    return emptyIconDisplay;
  };

  const filledIconDisplay = () => {
    const filledIconDisplay: VNode[] = [];
    for (let i = 0; i < props.numberOfTotalIcons; i++) {
      filledIconDisplay.push(renderIcon(true));
    }
    return filledIconDisplay;
  };

  return (
    <div
      class="relative w-max"
      part="value-rating"
      role="img"
      aria-label={props.i18n.t('stars', {
        count: props.numberOfActiveIcons,
        max: props.numberOfTotalIcons,
      })}
    >
      <div class="z-0 flex gap-0.5">{emptyIconDisplay()}</div>
      <div
        class="absolute left-0 top-0 z-1 flex gap-0.5 overflow-hidden"
        style={{width}}
      >
        {filledIconDisplay()}
      </div>
    </div>
  );
};

export const computeNumberOfStars = (
  value: unknown,
  field: string
): number | null => {
  if (value === null) {
    return null;
  }
  const valueAsNumber = parseFloat(`${value}`);
  if (Number.isNaN(valueAsNumber)) {
    throw new Error(
      `Could not parse "${value}" from field "${field}" as a number.`
    );
  }
  return valueAsNumber;
};
