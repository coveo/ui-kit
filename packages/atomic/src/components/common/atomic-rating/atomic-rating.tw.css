
@reference '../../../utils/tailwind.global.tw.css';
@reference '../../../utils/coveo.tw.css';
/**
 * @prop --atomic-rating-icon-outline: currentColor;line: Color of the icon's outline.
 */
 @layer components {
atomic-product-rating {
.text-rating-icon-active,
.text-rating-icon-inactive {
  & svg path {
    @apply stroke-rating-icon-outline;
  }
}}
 }