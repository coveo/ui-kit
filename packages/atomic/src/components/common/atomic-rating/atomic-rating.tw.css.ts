import {css} from 'lit';

const styles = css`
@reference '../../../utils/tailwind.global.tw.css';
/**
 * @prop --atomic-rating-icon-outline: Color of the icon's outline.
 */
@layer components {
  atomic-product-rating {
    .text-rating-icon-active,
    .text-rating-icon-inactive {
      & svg path {
        @apply stroke-rating-icon-outline;
      }
    }
  }
}
`;

export default styles;
