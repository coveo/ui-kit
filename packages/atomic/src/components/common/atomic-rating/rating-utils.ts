import {FieldValueIsNaNError} from '@/src/components/commerce/product-template-component-utils/error';

export const computeNumberOfStars = (
  value: unknown,
  field: string
): number | null => {
  if (value === null) {
    return null;
  }
  const valueAsNumber = parseFloat(`${value}`);
  if (Number.isNaN(valueAsNumber)) {
    throw new FieldValueIsNaNError(field, value);
  }
  return valueAsNumber;
};
