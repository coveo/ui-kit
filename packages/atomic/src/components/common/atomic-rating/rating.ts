import type {i18n} from 'i18next';
import {html} from 'lit';
import {styleMap} from 'lit/directives/style-map.js';
import {multiClassMap, tw} from '@/src/directives/multi-class-map';
import type {FunctionalComponent} from '@/src/utils/functional-component-utils';
import {FieldValueIsNaNError} from '../../commerce/product-template-component-utils/error';

interface RatingProps {
  i18n: i18n;
  numberOfTotalIcons: number;
  numberOfActiveIcons: number;
  icon: string;
  iconSize?: number;
}

export type {RatingProps};
/**
 * Renders a rating component.
 *
 * @part value-rating
 * @part value-rating-icon
 *
 * @cssprop --atomic-rating-icon-active-color - Color of the icon when active.
 * @cssprop --atomic-rating-icon-inactive-color - Color of the icon when inactive.
 * @cssprop --atomic-rating-icon-outline - Outline color of the icon.
 */
export const renderRating: FunctionalComponent<RatingProps> = ({props}) => {
  const width =
    ((props.numberOfActiveIcons / props.numberOfTotalIcons) * 100).toString() +
    '%';
  const iconSize = `${props.iconSize ?? 0.75}rem`;

  const renderIcon = (active: boolean) => {
    const iconClass = tw({
      'shrink-0': true,
      'text-rating-icon-active': active,
      'text-rating-icon-inactive': !active,
    });

    const iconStyle = {
      width: iconSize,
      height: iconSize,
      stroke: 'var(--atomic-rating-icon-outline)',
    };
    return html`<atomic-icon
      icon=${props.icon}
      class=${multiClassMap(iconClass)}
      style=${styleMap(iconStyle)}
      part="value-rating-icon"
    ></atomic-icon>`;
  };

  const emptyIconDisplay = () => {
    const emptyIcons = [];
    for (let i = 0; i < props.numberOfTotalIcons; i++) {
      emptyIcons.push(renderIcon(false));
    }
    return emptyIcons;
  };

  const filledIconDisplay = () => {
    const filledIcons = [];
    for (let i = 0; i < props.numberOfTotalIcons; i++) {
      filledIcons.push(renderIcon(true));
    }
    return filledIcons;
  };

  return html`<div
    class="relative w-max"
    part="value-rating"
    role="img"
    aria-label=${props.i18n.t('stars', {
      count: props.numberOfActiveIcons,
      max: props.numberOfTotalIcons,
    })}
  >
    <div class="z-0 flex gap-0.5">${emptyIconDisplay()}</div>
    <div
      class="absolute top-0 left-0 z-1 flex gap-0.5 overflow-hidden"
      style=${styleMap({width})}
    >
      ${filledIconDisplay()}
    </div>
  </div>`;
};

export const computeNumberOfStars = (
  value: unknown,
  field: string
): number | null => {
  if (value === null) {
    return null;
  }
  const valueAsNumber = parseFloat(`${value}`);
  if (Number.isNaN(valueAsNumber)) {
    throw new FieldValueIsNaNError(field, value);
  }
  return valueAsNumber;
};
