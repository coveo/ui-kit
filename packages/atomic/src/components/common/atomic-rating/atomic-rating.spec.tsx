import {computeNumberOfStars} from './atomic-rating';

describe('#computeNumberOfStars', () => {
  test('should return null when value is null', () => {
    const value = null;
    const field = 'rating';

    const result = computeNumberOfStars(value, field);

    expect(result).toBeNull();
  });

  test('should parse value as a number', () => {
    const value = '4.5';
    const field = 'rating';

    const result = computeNumberOfStars(value, field);

    expect(result).toBe(4.5);
  });

  test('should throw an error when value is not a number', () => {
    const value = 'invalid';
    const field = 'rating';

    expect(() => {
      computeNumberOfStars(value, field);
    }).toThrowErrorMatchingSnapshot();
  });
});
