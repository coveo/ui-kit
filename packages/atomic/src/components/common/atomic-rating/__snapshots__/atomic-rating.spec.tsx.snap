// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#computeNumberOfStars should throw an error when value is not a number 1`] = `"Could not parse "invalid" from field "rating" as a number."`;
